dspy/primitives/program.py

```diff
@@ -1,9 +1,195 @@
+from __future__ import annotations  # Used for self type hints
+from typing import Optional
+
 import magicattr
 
+from dsp.modules.lm import LM
+import dspy
 from dspy.primitives.assertions import *
 from dspy.primitives.module import BaseModule
 
 
+#-------------------------------------------------------------------------------
+#    Templates for the user-facing strings used by this module
+#-------------------------------------------------------------------------------
+
+_ERR_MSG_NUM_PREDICTORS = """Structurally equivalent programs must have the \
+the number of predictors. The number of predictors for the two modules do not \
+match: {num1} != {num2}"""
+
+_ERR_MSG_PREDICTOR_NAMES = """Program predictor names must match at \
+corresponding indices for structural equivalency. The predictor names for the \
+programs do not match at index {ind}: '{name1}' != '{name2}'"""
+
+_ERR_MSG_SHARED_PREDICTORS = """The programs share the following predictor(s) \
+with each other: {pred_names}"""
+
+_ERR_MSG_UNSET_DSPY_LM = """LM consistency violated: Expected all predictors’ \
+LMs to be set when dspy.settings.lm is set to ‘NoneType’, but some \
+predictors have 'NoneType' LMs.
+
+{lm_info}"""
+
+_ERR_MSG_SET_DSPY_LM = """LM consistency violated: Expected all predictors' \
+LMs to be 'NoneType' when 'dspy.settings.lm' is set, but some predictors have \
+LMs set.
+
+{lm_info}"""
+
+_ERR_MSG_PROGRAM_LM = """Setting or getting the LM of a program is an \
+experimental feature. Please enable the 'dspy.settings.experimental' flag to \
+use these feature."""
+
+_INFO_LM_PROGRAM = """The LM set in 'dspy.settings.lm' is an instance of \
+'{lname}' LMs set for the predictors in the module are:{pred_lm_info}"""
+
+_INFO_LM_PRED = """
+    LM for {pname} is an instance of ’{lname}’"""
+
+
+#-------------------------------------------------------------------------------
+#    Helper functions
+#-------------------------------------------------------------------------------
+
+def assert_structural_equivalency_for_programs(
```

Comment:
These functions are cluttering this file. Do we even need them? If yes, do they need to be in this file? They seem like nice utils to have in a different file.

The code "smell" here that gives away that they don't need to be here is that the author decided to have the divisioning below, which we don't really like to have in DSPy:

```python
-------------------------------------------------------------------------------
#    Helper functions
#-------------------------------------------------------------------------------
```

Reply:
I think primitives/utils.py and primitives/constants.py can be good homes for these or just all in primitives/utils.py

Reply:
`primitives/utils.py` is the right location.

The type hints is wrong here as well.

Reply:
(Looks like we moved them to the teleprompter file, resolving)